Modules/_zoneinfo.c

```diff
@@ -2203,7 +2205,8 @@ find_ttinfo(zoneinfo_state *state, PyZoneInfo_ZoneInfo *self, PyObject *dt)
         return NULL;
     }
 
-    unsigned char fold = PyDateTime_DATE_GET_FOLD(dt);
+    PyDateTime_DateTime *dt_dt = as_datetime(dt);
```

Comment:
**issue:** This function is now on the hook to deallocate `dt_dt`. So in all return paths we must `Py_DECREF(dt_dt)`. Since this function has multiple points where it returns a value that either means refactoring to assign to a result and handling deallocation that way, or, perhaps cleaner:

**suggestion:** Since with this change `get_local_timestamp` _also_ calls `as_datetime` to parse out the timestamp, how about we push the obligation to deliver a `PyDateTime_Datetime *` onto the _caller_ of `find_ttinfo`? So then, for example, `zoneinfo_ZoneInfo_utcoffset_impl` would be:

```
static PyObject *
zoneinfo_ZoneInfo_utcoffset_impl(PyObject *self, PyTypeObject *cls,
                                 PyObject *dt)
/*[clinic end generated code: output=b71016c319ba1f91 input=2bb6c5364938f19c]*/
{
    zoneinfo_state *state = zoneinfo_get_state_by_cls(cls);
    PyDateTime_DateTime *dtdt = as_datetime(dt);
    _ttinfo *tti = find_ttinfo(state, (PyZoneInfo_ZoneInfo *)self, dtdt);
    Py_DECREF(dtdt);
    if (tti == NULL) {
        return NULL;
    }
    return Py_NewRef(tti->utcoff);
}
```
`find_ttinfo` remains unchanged from its implementation on trunk (except the `dt` argument becomes a `PyDateTime_DateTime *`). Similarly `get_local_timestamp` can then narrow its input type to `PyDateTime_DateTime *`.

Reply:
I began with this approach but couldn't pass the tests for zoneinfo. My feeling was I had some issue with the reference counting